engine/table/src/main/java/io/deephaven/engine/table/impl/ZeroKeyUpdateBy.java

```diff
@@ -59,469 +68,271 @@ public static Table compute(@NotNull final String description,
     }
 
+    private void processUpdateForSsa(TableUpdate upstream) {
+        final boolean stampModified = upstream.modifiedColumnSet().containsAny(timestampColumnSet);
+
+        final RowSet restampRemovals;
+        final RowSet restampAdditions;
+
+        // modifies are remove + add operations
+        if (stampModified) {
+            restampAdditions = upstream.added().union(upstream.modified());
+            restampRemovals = upstream.removed().union(upstream.getModifiedPreShift());
+        } else {
+            restampAdditions = upstream.added();
+            restampRemovals = upstream.removed();
+        }
+
+        // removes
+        if (restampRemovals.isNonempty()) {
+            final int size = (int) Math.min(restampRemovals.size(), 4096);
+            try (final RowSequence.Iterator it = restampRemovals.getRowSequenceIterator();
+                    final ChunkSource.GetContext context = timestampColumnSource.makeGetContext(size);
+                    final WritableLongChunk<? extends Values> ssaValues = WritableLongChunk.makeWritableChunk(size);
+                    final WritableLongChunk<OrderedRowKeys> ssaKeys = WritableLongChunk.makeWritableChunk(size)) {
+
+                MutableLong lastTimestamp = new MutableLong(NULL_LONG);
+                while (it.hasMore()) {
+                    RowSequence chunkRs = it.getNextRowSequenceWithLength(4096);
+
+                    // get the chunks for values and keys
+                    LongChunk<? extends Values> valuesChunk =
+                            timestampColumnSource.getPrevChunk(context, chunkRs).asLongChunk();
+                    LongChunk<OrderedRowKeys> keysChunk = chunkRs.asRowKeyChunk();
+
+                    // push only non-null values/keys into the Ssa
+                    fillChunkWithNonNull(keysChunk, valuesChunk, ssaKeys, ssaValues, lastTimestamp);
+                    timestampSsa.remove(ssaValues, ssaKeys);
+                }
+            }
+        }
+
+        // shifts
+        if (upstream.shifted().nonempty()) {
+            final int size = Math.max(
+                    upstream.modified().intSize() + Math.max(upstream.added().intSize(), upstream.removed().intSize()),
+                    (int) upstream.shifted().getEffectiveSize());
+            try (final RowSet fullPrevRowSet = source.getRowSet().copyPrev();
+                    final RowSet previousToShift = fullPrevRowSet.minus(restampRemovals);
+                    final ColumnSource.GetContext getContext = timestampColumnSource.makeGetContext(size)) {
+
+                final RowSetShiftData.Iterator sit = upstream.shifted().applyIterator();
+                while (sit.hasNext()) {
+                    sit.next();
+                    try (final RowSet subRowSet = previousToShift.subSetByKeyRange(sit.beginRange(), sit.endRange());
+                            final RowSet rowSetToShift = subRowSet.minus(upstream.removed())) {
```

Comment:
It seems like it would be better to do that removal one time rather than every time.